### PR TITLE
#31 Add reconciler for Policies

### PR DESCRIPTION
**Pull request by bfink13**

### CODE REVIEW

File: rabbitmq-operator-model/src/main/java/com/indeed/operators/rabbitmq/model/crd/rabbitmq/PolicyDefinitionSpec.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    public Long getMessageTtl() {
+        return messageTtl;
+    }
+
+    public String getQueueMasterLocator() {
+        return queueMasterLocator;
+    }
+
+    @JsonIgnore
+    public Definition asDefinition() {
```
Comment (bfink13) [inline rabbitmq-operator-model/src/main/java/com/indeed/operators/rabbitmq/model/crd/rabbitmq/PolicyDefinitionSpec.java:106]: this is sort of terrible...i'm open to suggestions

File: rabbitmq-operator-model/src/main/java/com/indeed/operators/rabbitmq/model/crd/rabbitmq/PolicyDefinitionSpec.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    private final Long haParams;
+    private final String haPromoteOnShutdown;
+    private final Long haSyncBatchSize;
+    private final String haSyncMode;
+    private final Long maxLength;
+    private final Long maxLengthBytes;
+    private final Long messageTtl;
+    private final String queueMasterLocator;
+
+    public PolicyDefinitionSpec(
```
Comment (bfink13) [inline rabbitmq-operator-model/src/main/java/com/indeed/operators/rabbitmq/model/crd/rabbitmq/PolicyDefinitionSpec.java:23]: @amccormick-indeed what are your thoughts about having these "spec" classes? Unfortunately this one is almost a direct copy of the rabbitmq-admin Definition class, but I feel like having the separation between the two realms (k8s and rabbit) is a good thing despite the duplicated code...

File: rabbitmq-operator/src/main/java/com/indeed/operators/rabbitmq/reconciliation/rabbitmq/PolicyReconciler.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import com.indeed.rabbitmq.admin.pojo.Policy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class PolicyReconciler {
```
Comment (bfink13) [inline rabbitmq-operator/src/main/java/com/indeed/operators/rabbitmq/reconciliation/rabbitmq/PolicyReconciler.java:18]: This is essentially a direct copy of the ShovelReconciler

File: rabbitmq-operator-model/src/main/java/com/indeed/operators/rabbitmq/model/crd/rabbitmq/PolicyDefinitionSpec.java (hunk 2 above)
Comment (ghost) [reply rabbitmq-operator-model/src/main/java/com/indeed/operators/rabbitmq/model/crd/rabbitmq/PolicyDefinitionSpec.java:23]: I agree with your thinking.

File: rabbitmq-operator-model/src/main/java/com/indeed/operators/rabbitmq/model/crd/rabbitmq/PolicyDefinitionSpec.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    public String getQueueMasterLocator() {
+        return queueMasterLocator;
+    }
+
+    @JsonIgnore
+    public Definition asDefinition() {
+        Definition definition = new Definition();
+
+        if (getAlternateExchange() != null) {
+            definition = definition.withAlternateExchange(getAlternateExchange());
```
Comment (ghost) [inline rabbitmq-operator-model/src/main/java/com/indeed/operators/rabbitmq/model/crd/rabbitmq/PolicyDefinitionSpec.java:110]: Can you pass `null` to `definition.withAlternateExchange`?  So you could avoid the null check at least?

**Review by ghost — APPROVED** [review]

File: rabbitmq-operator-model/src/main/java/com/indeed/operators/rabbitmq/model/crd/rabbitmq/PolicyDefinitionSpec.java (hunk 4 above)
Comment (bfink13) [reply rabbitmq-operator-model/src/main/java/com/indeed/operators/rabbitmq/model/crd/rabbitmq/PolicyDefinitionSpec.java:110]: unfortunately it seems the "null"s get serialized in the request payload. we could probably update rabbitmq-admin to not do that. or at least provide the option.

or EVEN BETTER, make the api factory non-static and supply our own ObjectMapper.

**Pull request merged**